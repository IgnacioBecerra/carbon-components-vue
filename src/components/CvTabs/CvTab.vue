<template>
  <div
    :class="`cv-tab ${carbonPrefix}--tab-content`"
    :id="uid"
    role="tabpanel"
    :aria-labelledby="`${uid}-link`"
    :aria-hidden="!dataSelected ? 'true' : 'false'"
    :hidden="!dataSelected"
  >
    <slot>
      <!-- Content for tab goes here. -->
    </slot>
  </div>
</template>

<script setup>
import { carbonPrefix } from '../../global/settings';
import { useCvId, props as propsCvId } from '../../use/cvId';
import { inject, onBeforeUnmount, onMounted, ref, watch } from 'vue';

const props = defineProps({
  selected: { type: Boolean, default: false },
  disabled: { type: Boolean, default: false },
  label: { type: String, required: true },
  ...propsCvId,
});
const uid = useCvId(props, true);

/**
 * Is this tab selected?
 * @type {Ref<UnwrapRef<string>>}
 */
const tabSelected = inject('tab-selected');
const dataSelected = ref(props.selected);
onMounted(() => {
  if (props.selected) tabSelected.value = uid.value;
});
watch(
  () => props.selected,
  () => {
    if (props.selected) tabSelected.value = uid.value;
  }
);
watch(tabSelected, () => {
  dataSelected.value = tabSelected.value === uid.value;
});

/**
 * Is this tab disabled?
 * @type {Ref<UnwrapRef<Set>>}
 */
const tabDisabledIds = inject('tab-disabled-ids');
function updateTabDisabled() {
  if (props.disabled) tabDisabledIds.value?.add(uid.value);
  else tabDisabledIds.value?.delete(uid.value);
}
onMounted(updateTabDisabled);
watch(() => props.disabled, updateTabDisabled);

/**
 * Keep this tab's data up to date
 * @type {Ref<UnwrapRef<Array<TabData>>>}
 */
const tabData = inject('tab-data');
function updateTabData() {
  const index = tabData.value?.findIndex(tab => tab.uid === uid.value);
  if (index > -1)
    tabData.value.splice(index, 1, { uid: uid.value, label: props.label });
  else tabData.value.push({ uid: uid.value, label: props.label });
}
onMounted(updateTabData);
watch(() => props.label, updateTabData);
onBeforeUnmount(() => {
  const index = tabData.value?.findIndex(tab => tab.uid === uid.value);
  if (index > -1) tabData.value.splice(index, 1);
});
</script>
