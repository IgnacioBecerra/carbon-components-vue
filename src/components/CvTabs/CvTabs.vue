<!--suppress ES6PreferShortImport -->
<template>
  <div class="cv-tabs" style="width: 100%">
    <div
      data-tabs
      :class="[
        `cv-tab ${carbonPrefix}--tabs--scrollable`,
        {
          [`${carbonPrefix}--tabs--container`]: container,
          [`${carbonPrefix}--tabs--scrollable--container`]: container,
        },
      ]"
      role="navigation"
      v-bind="$attrs"
      @keydown.right.prevent.stop="onRight"
      @keydown.left.prevent.stop="onLeft"
    >
      <button
        aria-hidden="true"
        aria-label="scroll left"
        :class="[
          {
            [`${carbonPrefix}--tab--overflow-nav-button`]: horizontalOverflow,
            [`${carbonPrefix}--tab--overflow-nav-button--hidden`]:
              leftOverflowNavButtonHidden,
          },
        ]"
        @click.stop.prevent="e => onOverflowClick(e, { direction: -1 })"
        @mousedown.stop.prevent="e => onOverflowMouseDown(e, { direction: -1 })"
        @mouseup.stop.prevent="onOverflowMouseUp"
        tabIndex="-1"
        type="button"
        ref="elLeftOverflow"
      >
        <ChevronLeft16 />
      </button>
      <div
        v-if="!leftOverflowNavButtonHidden"
        :class="`${carbonPrefix}--tabs__overflow-indicator--left`"
      />

      <ul
        :class="`${carbonPrefix}--tabs--scrollable__nav`"
        role="tablist"
        ref="elTabList"
      >
        <li
          v-for="tab in tabs"
          :key="tab.uid"
          :class="[
            `cv-tabs-button  ${carbonPrefix}--tabs--scrollable__nav-item`,
            {
              [`${carbonPrefix}--tabs__nav-item--selected`]:
                selectedId === tab.uid,
              [`${carbonPrefix}--tabs__nav-item--disabled`]: disabledTabs.has(
                tab.uid
              ),
              // TODO: remove scrollable in next major release
              [`${carbonPrefix}--tabs--scrollable__nav-item--disabled`]:
                disabledTabs.has(tab.uid),
              [`${carbonPrefix}--tabs--scrollable__nav-item--selected`]:
                selectedId === tab.uid,
            },
          ]"
          role="presentation"
        >
          <button
            :class="`${carbonPrefix}--tabs--scrollable__nav-link`"
            role="tab"
            :aria-controls="tab.uid"
            :aria-disabled="disabledTabs.has(tab.uid)"
            :aria-selected="selectedId === tab.uid"
            :id="`${tab.uid}-link`"
            @click="onTabClick(tab.uid)"
            ref="elLink"
            :tabindex="selectedId === tab.uid ? 0 : -1"
            type="button"
          >
            {{ tab.label }}
          </button>
        </li>
      </ul>

      <div
        v-if="!rightOverflowNavButtonHidden"
        :class="`${carbonPrefix}--tabs__overflow-indicator--right`"
      />
      <button
        aria-hidden="true"
        aria-label="scroll right"
        :class="[
          {
            [`${carbonPrefix}--tab--overflow-nav-button`]: horizontalOverflow,
            [`${carbonPrefix}--tab--overflow-nav-button--hidden`]:
              rightOverflowNavButtonHidden,
          },
        ]"
        @click="e => onOverflowClick(e, { direction: 1 })"
        @mousedown="e => onOverflowMouseDown(e, { direction: 1 })"
        @mouseup="onOverflowMouseUp"
        tabIndex="-1"
        type="button"
        ref="elRightOverflow"
      >
        <ChevronRight16 />
      </button>
    </div>
    <div class="cv-tabs__panels">
      <slot></slot>
    </div>
  </div>
</template>

<script setup>
import { carbonPrefix } from '../../global/settings';
import { ChevronLeft16, ChevronRight16 } from '@carbon/icons-vue';
import {
  nextTick,
  onBeforeUnmount,
  onMounted,
  onUpdated,
  provide,
  ref,
  watch,
} from 'vue';

const OVERFLOW_BUTTON_OFFSET = 40;

const props = defineProps({
  /**
   * Changes the look of the tabs themselves. Does NOT change the styles of the tab content.
   */
  container: { type: Boolean, default: false },
  /**
   * @deprecated not used in component
   */
  leftOverflowIconButtonProps: Object,
  /**
   * This is an edge use case. By default, if no tab is explicitly selected, the first non-disabled tab is selected.
   */
  noDefaultToFirst: { type: Boolean, default: false },
  /**
   * @deprecated not used in component
   */
  rightOverflowIconButtonProps: Object,
  /**
   * @deprecated not used in component
   */
  scrollIntoView: { type: Boolean, default: true },
});
const emit = defineEmits(['tab-selected', 'tab-selected-id']);

const selectedId = ref('');
provide('tab-selected', selectedId);
/**
 * @typedef TabData
 * @type {string} uid
 * @type {string} label
 */

/**
 * @type {Ref<UnwrapRef<Array<TabData>>>}
 */
const tabs = ref([]);
provide('tab-data', tabs);
const disabledTabs = ref(new Set());
provide('tab-disabled-ids', disabledTabs);
const horizontalOverflow = ref(false);
const leftOverflowNavButtonHidden = ref(false);
const rightOverflowNavButtonHidden = ref(false);

function maybeSelectFirstTab() {
  // Maybe user does not want something selected by default? Seems like this is an odd use case.
  if (props.noDefaultToFirst) return;

  // is something selected?
  if (tabs.value.find(tab => tab.uid === selectedId.value)) return;

  // Find the first not-disabled tab
  const index = tabs.value?.findIndex(
    tabData => !disabledTabs.value?.has(tabData.uid)
  );

  // if there is a non-disabled tab set that one as selected
  if (index > -1) selectedId.value = tabs.value[index].uid;
}
onMounted(maybeSelectFirstTab);
watch(tabs, maybeSelectFirstTab);
watch(selectedId, () => {
  nextTick(() => {
    doTabClick(selectedId.value);
  });
});
const elTabList = ref(null);
onMounted(() => {
  window.addEventListener('resize', handleResize);
  elTabList.value?.addEventListener('scroll', checkScroll);
  checkScroll();
});

onUpdated(() => {
  checkScroll();
});
onBeforeUnmount(() => {
  window.removeEventListener('resize', handleResize);
  elTabList.value?.removeEventListener('scroll', checkScroll);
});

const skipScroll = ref(false);
const elLink = ref(null);
function checkScroll() {
  if (!skipScroll.value) {
    if (elTabList.value) {
      horizontalOverflow.value =
        elTabList.value?.scrollWidth > elTabList.value?.clientWidth;

      if (elLink.value && elLink.value.length > 0) {
        leftOverflowNavButtonHidden.value = elTabList.value?.scrollLeft <= 0;
        rightOverflowNavButtonHidden.value =
          elTabList.value?.scrollLeft + elTabList.value?.clientWidth >=
          elTabList.value?.scrollWidth;
      }
    }
  }
}
function handleResize() {
  checkScroll();
}

function doScrollIntoView(index) {
  const tab = elLink.value[index];
  /**
   * HTMLElement
   * @type {HTMLElement|null}
   */
  const scrollContainer =
    tab && tab.parentNode ? tab.parentNode.parentNode : null;
  let newScrollLeft;

  if (tab && scrollContainer) {
    const scrollLeft = scrollContainer.scrollLeft;
    const tabLeft = tab.offsetLeft - scrollContainer.offsetLeft;

    if (index === 0) {
      newScrollLeft = 0;
    } else if (index === elLink.value.length - 1) {
      newScrollLeft = scrollContainer.scrollWidth - scrollContainer.offsetWidth;
    } else if (tabLeft < scrollLeft) {
      newScrollLeft = tabLeft;
    } else {
      const rightOfTab =
        tab.offsetLeft - scrollContainer.offsetLeft + tab.offsetWidth;
      if (rightOfTab > scrollContainer.offsetWidth + scrollLeft) {
        newScrollLeft = rightOfTab - scrollContainer.offsetWidth;

        if (!rightOverflowNavButtonHidden.value) {
          newScrollLeft += OVERFLOW_BUTTON_OFFSET;
        }
      }
    }
  }

  if (newScrollLeft !== undefined) {
    skipScroll.value = true;
    leftOverflowNavButtonHidden.value = newScrollLeft <= 0;
    rightOverflowNavButtonHidden.value =
      newScrollLeft + elTabList.value?.clientWidth >=
      elTabList.value?.scrollWidth;

    nextTick(() => {
      // allow left and right nav hide to propagate before setting scroll
      scrollContainer.scrollLeft = newScrollLeft;
      skipScroll.value = false;
    });
  }
}
function doTabClick(id, setFocus = false) {
  if (!disabledTabs.value.has(id)) {
    if (selectedId.value !== id) {
      let newIndex = -1;

      selectedId.value = id;

      for (let i = 0; i < tabs.value.length; i++) {
        if (tabs.value[i].uid === id) {
          doScrollIntoView(i, setFocus);
          if (setFocus) {
            nextTick(() => {
              elLink.value[i].focus();
            });
          }
          newIndex = i;
        }
      }

      emit('tab-selected', newIndex);
      emit('tab-selected-id', tabs.value[newIndex].uid);
    }
  }
}
function onTabClick(id) {
  doTabClick(id);
}
function isAllTabsDisabled() {
  return disabledTabs.value.size === tabs.value.length;
}
function onLeft() {
  if (isAllTabsDisabled()) {
    return;
  }

  const curIndex = move(selectedId.value, false);
  const newId = tabs.value[curIndex].uid;

  doTabClick(newId, true);
}
function move(id, next) {
  let newIndex;
  let newId;

  newIndex = tabs.value.findIndex(tab => tab.uid === id);

  if (newIndex > -1) {
    newIndex = next ? newIndex + 1 : newIndex - 1;
  }
  if (newIndex < 0 || newIndex >= tabs.value.length) {
    newIndex = next ? 0 : tabs.value.length - 1;
  }

  newId = tabs.value[newIndex].uid;
  while (newId === selectedId.value || disabledTabs.value.has(newId)) {
    if (newIndex > -1) {
      newIndex = next ? newIndex + 1 : newIndex - 1;
    }
    if (newIndex < 0 || newIndex >= tabs.value.length) {
      newIndex = next ? 0 : tabs.value.length - 1;
    }
    newId = tabs.value[newIndex].uid;
  }

  return newIndex;
}
function onRight() {
  if (isAllTabsDisabled()) {
    return;
  }

  const curIndex = move(selectedId.value, true);
  const newId = tabs.value[curIndex].uid;

  doTabClick(newId, true);
}
const elRightOverflow = ref(null);
const elLeftOverflow = ref(null);
function onOverflowClick(e, { direction, multiplier = 10 }) {
  const scrollLeft = elTabList.value?.scrollLeft;

  // account for overflow button appearing and causing tablist width change
  if (elTabList.value && direction === 1 && !scrollLeft) {
    elTabList.value.scrollLeft += OVERFLOW_BUTTON_OFFSET;
  }

  if (elTabList.value) elTabList.value.scrollLeft += direction * multiplier;

  if (elTabList.value && leftEdgeReached(direction)) {
    elTabList.value.scrollLeft = 0;
  }

  // account for reaching left edge
  if (leftEdgeReached(direction)) {
    elRightOverflow.value?.focus();
  }
  if (rightEdgeReached(direction)) {
    elLeftOverflow.value?.focus();
  }
}
let overflowNavInterval;
function onOverflowMouseDown(e, { direction }) {
  // disregard mouse buttons aside from LMB
  if (e.buttons !== 1) {
    return;
  }

  if (overflowNavInterval) clearInterval(overflowNavInterval);
  overflowNavInterval = setInterval(() => {
    if (leftEdgeReached(direction) || rightEdgeReached(direction)) {
      clearInterval(overflowNavInterval);
    }

    onOverflowClick(e, { direction });
  });
}
function onOverflowMouseUp() {
  clearInterval(overflowNavInterval);
}
function leftEdgeReached(direction) {
  const scrollLeft = elTabList.value?.scrollLeft || Number.MIN_VALUE;

  return direction === -1 && scrollLeft <= OVERFLOW_BUTTON_OFFSET;
}
function rightEdgeReached(direction) {
  if (!elTabList.value) return false;
  const { clientWidth, scrollLeft, scrollWidth } = elTabList.value;
  return direction === 1 && scrollLeft + clientWidth >= scrollWidth;
}
</script>
