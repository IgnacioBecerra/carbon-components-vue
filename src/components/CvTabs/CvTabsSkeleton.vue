<!--suppress ALL -->
<template>
  <div
    :class="`cv-tabs ${carbonPrefix}--tabs ${carbonPrefix}--skeleton ${carbonPrefix}--tabs--scrollable`"
  >
    <ul :class="`${carbonPrefix}--tabs--scrollable__nav`">
      <li
        v-for="index in tabCount"
        :key="index"
        :class="`${carbonPrefix}--tabs--scrollable__nav-item`"
      >
        <div :class="`${carbonPrefix}--tabs__nav-link`"><span></span></div>
      </li>
    </ul>
  </div>
</template>

<script setup>
import { carbonPrefix } from '../../global/settings';

defineProps({
  tabCount: { type: Number, default: 5 },
});
</script>
