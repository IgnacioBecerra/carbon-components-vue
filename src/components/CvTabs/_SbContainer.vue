<template>
  <div class="sb-tabs-container">
    <slot></slot>
  </div>
</template>

<script setup></script>

<style>
.bx--tabs--container ~ div {
  min-height: 320px;
  background-color: #e8daff;
  background-repeat: no-repeat;
  background-position: center;
}
.bx--tabs--container ~ .cv-tabs__panels:after {
  content: 'Container style is not from carbon. This example uses ".bx--tabs--container ~ div" selector to add the "#e8daff" background color.';
  opacity: 0.3;
  font-size: 30pt;
  z-index: -10;
}
</style>
