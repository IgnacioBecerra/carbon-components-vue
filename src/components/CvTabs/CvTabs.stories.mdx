import { Canvas, Meta, Story } from '@storybook/addon-docs';
import { sbCompPrefix } from '../../global/storybook-utils';
import CvTabs from './CvTabs.vue';
import CvTab from './CvTab.vue';
import CvTabsSkeleton from './CvTabsSkeleton.vue';
import SbContainer from './_SbContainer.vue';
import { action } from '@storybook/addon-actions';

<Meta
  title={`${sbCompPrefix}/CvTabs`}
  component={CvTabs}
  decorators={[
    () => ({
      components: { SbContainer, Story },
      template: `<SbContainer><story/></SbContainer>`,
    }),
  ]}
/>

export const Template = args => ({
  // Components used in your story `template` are defined in the `components` object
  components: {
    CvTabs,
    CvTab,
    CvTabsSkeleton,
  },
  // The story's `args` need to be mapped into the template through the `setup()` method
  setup() {
    return {
      container: args.container,
      noDefaultToFirst: args.noDefaultToFirst,
      selectedId: args.selectedId,
      disabledIds: args.disabledIds,
      onTabSelected: action('tab-selected'),
      onTabSelectedId: action('tab-selected-id'),
    };
  },
  // And then the `args` are bound to your component with `v-bind="args"`
  template: args.template,
});
const defaultTemplate = `
<cv-tabs
  :container="container"
  :no-default-to-first="noDefaultToFirst"
  @tab-selected="onTabSelected"
  @tab-selected-id="onTabSelectedId"
  aria-label="navigation tab label">
    <cv-tab id="tab-1" label="House" :selected="selectedId === 'tab-1'" :disabled="disabledIds.includes('tab-1')">
      <p>
        29 Arlington Avenue is a house in Islington, London. Douglas Adams shared the house for a time
        with Jonny Brock and Clare Gorst. It is mentioned in the dedication of The Hitchhiker's Guide to
        the Galaxy, which is dedicated to the "Arlingtonians for tea, sympathy, and a sofa."
      </p>
    </cv-tab>
    <cv-tab id="tab-2" label="Bar" :selected="selectedId === 'tab-2'" :disabled="disabledIds.includes('tab-2')">
      <p>
        The Old Pink Dog Bar was the bar in Han Dold City that Ford Prefect was in when he discovered
        that his Guide had his full entry on Earth on it.
      </p>
    </cv-tab>
    <cv-tab id="tab-3" label="Reach" :selected="selectedId === 'tab-3'" :disabled="disabledIds.includes('tab-3')">
      <p>
        The Third Reach of the Unknown is what might be loosely described as a "place" which exists only
        under high improbability conditions and was created on the starship Heart of Gold due to the
        Infinite Improbability Drive. During Arthur Dent and Ford Prefect's first trip on board the Heart
        of Gold they witnessed many strange things happening.
      </p>
    </cv-tab>
    <cv-tab id="tab-4" label="Zone" :selected="selectedId === 'tab-4'" :disabled="disabledIds.includes('tab-4')">
      <p>Plural Zones are instabilities in space-time.
        The effects of the Plural Zones on the beings and objects that inhabit and originate from them,
        for the most part, remain largely unnoticed.
      </p>
    </cv-tab>
    <cv-tab id="tab-5" label="Planet" :selected="selectedId === 'tab-5'" :disabled="disabledIds.includes('tab-5')">
      <p>
        Ursa Minor Beta is a rich and sunny planet and one of the most popular holiday destinations in
        the Universe. It is always Saturday afternoon on this planet, just before the beach bars close -
        except for the few places where it is eternally Saturday evening. This is why the locals always
        invite visitors to "Have a nice diurnal anomaly!"
        <a target='_blank' href='https://hitchhikers.fandom.com/wiki/Main_Page'>credits</a>
      </p>
    </cv-tab>
</cv-tabs>
`;
const skeletonTemplate = `<cv-tabs-skeleton></cv-tabs-skeleton>`;

# CvTabs

Migration notes:

- The property `leftOverflowIconButtonProps` is deprecated. It does not seem to be used at all in Vue 2 and thus does
  nothing when specified.
- The property `rightOverflowIconButtonProps` is deprecated. It does not seem to be used at all in Vue 2 and thus does
  nothing when specified.
- The property `scrollIntoView` is deprecated. It does not seem to be used at all in Vue 2 and thus does nothing when
  specified.
- The `container` property in Vue 2 acts differently as compared to the React story book. Here the class/styles are
  applied as is shown in the React storybook. Also note that the background color applied when this property is
  specified is not from Carbon. The selector `.bx--tabs--container ~ div` can be used to set the background color and
  other properties but that is up to the user.
- A new event emits `tab-selected-id` with the selected tab's id attribute.

## Mobile

In my opinion, this component is not suitable for mobile formats. The navigation of the tabs becomes awkward on smaller
screens, and so I suggest using a different component for mobile cases.

<Canvas>
  <Story
    name="Default"
    parameters={{
      controls: {
        exclude: [
          'default',
          'template',
          'leftOverflowIconButtonProps',
          'rightOverflowIconButtonProps',
          'scrollIntoView',
          'tab-selected',
          'tab-selected-id',
        ],
      },
      docs: { source: { code: defaultTemplate } },
    }}
    args={{
      template: defaultTemplate,
      selectedId: '',
      disabledIds: [],
    }}
    argTypes={{
      selectedId: {
        control: 'select',
        default: '',
        options: ['', 'tab-1', 'tab-2', 'tab-3', 'tab-4', 'tab-5'],
      },
      disabledIds: {
        control: 'multi-select',
        options: ['tab-1', 'tab-2', 'tab-3', 'tab-4', 'tab-5'],
      },
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

# Skeleton

<Canvas>
  <Story
    name="skeleton"
    parameters={{
      controls: {
        exclude: [
          'default',
          'template',
          'leftOverflowIconButtonProps',
          'rightOverflowIconButtonProps',
          'scrollIntoView',
          'tab-selected',
          'tab-selected-id',
          'container',
          'noDefaultToFirst',
        ],
      },
      docs: { source: { code: skeletonTemplate } },
    }}
    args={{
      template: skeletonTemplate,
    }}
    argTypes={{}}
  >
    {Template.bind({})}
  </Story>
</Canvas>
