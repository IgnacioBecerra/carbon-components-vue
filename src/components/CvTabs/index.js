import CvTabs from './CvTabs.vue';
import CvTab from './CvTab.vue';
import CvTabsSkeleton from './CvTabsSkeleton.vue';

export { CvTab, CvTabs, CvTabsSkeleton };
