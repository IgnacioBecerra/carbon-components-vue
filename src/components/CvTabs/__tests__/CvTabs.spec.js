import { render } from '@testing-library/vue';
import { mount } from '@vue/test-utils';
import userEvent from '@testing-library/user-event';
import CvTabs from '../CvTabs.vue';
import CvTab from '../CvTab.vue';

const Tab1Label = 'Tab 1 label';
const Tab2Label = 'Tab 2 label';
const ariaLabel = 'ABC-aria-label-123';

const ContentTabs = {
  components: { CvTab },
  template: `
    <cv-tab  id="tab-1" label="${Tab1Label}">Tab 1 content</cv-tab>
    <cv-tab  id="tab-2" label="${Tab2Label}">Tab 2 content</cv-tab>
    <cv-tab  id="tab-3" label="Tab 3 label">Tab 3 content</cv-tab>
  `,
};

describe('CvTabs', () => {
  it('CvTabs - test default and attrs', async () => {
    // The render method returns a collection of utilities to query your component.
    const result = render(CvTabs, {
      props: {},
      slots: {
        default: ContentTabs,
      },
      attrs: {
        class: 'ABC-class-123',
        'aria-label': ariaLabel,
      },
    });
    const user = userEvent.setup();

    const tabNav = await result.findByRole('navigation');
    const allTabs = await result.findAllByRole('tab');
    expect(allTabs.length).toBe(3);

    // check content
    const content1 = await result.findByText('Tab 1 content');
    const content2 = await result.findByText('Tab 2 content');
    const content3 = await result.findByText('Tab 3 content');
    expect(content1.getAttribute('aria-hidden')).toBe('false');
    expect(content2.getAttribute('aria-hidden')).toBe('true');
    expect(content3.getAttribute('aria-hidden')).toBe('true');

    await user.click(allTabs[1]);
    expect(content1.getAttribute('aria-hidden')).toBe('true');
    expect(content2.getAttribute('aria-hidden')).toBe('false');
    expect(content3.getAttribute('aria-hidden')).toBe('true');

    await user.click(allTabs[2]);
    expect(content1.getAttribute('aria-hidden')).toBe('true');
    expect(content2.getAttribute('aria-hidden')).toBe('true');
    expect(content3.getAttribute('aria-hidden')).toBe('false');

    expect(tabNav.classList.contains('ABC-class-123')).toBe(true);
    expect(tabNav.getAttribute('aria-label')).toBe(ariaLabel);
    expect(result.emitted('tab-selected').length).toBe(2);
    expect(result.emitted('tab-selected')[0][0]).toBe(1);
    expect(result.emitted('tab-selected-id').length).toBe(2);
    expect(result.emitted('tab-selected-id')[0][0]).toBe('tab-2');
  });

  it('CvTabs - test all props', async () => {
    const props = {
      id: 'ABC-123',
      container: false,
    };
    // The render method returns a collection of utilities to query your component.
    const result = render(CvTabs, {
      props,
      slots: {
        default: ContentTabs,
      },
      attrs: {
        'aria-label': ariaLabel,
      },
    });

    const tabNav = await result.findByRole('navigation');
    expect(tabNav.classList.contains('bx--tabs--container')).toBe(false);
    expect(tabNav.classList.contains('bx--tabs--scrollable--container')).toBe(
      false
    );
    await result.rerender({ container: true });
    expect(tabNav.classList.contains('bx--tabs--container')).toBe(true);
    expect(tabNav.classList.contains('bx--tabs--scrollable--container')).toBe(
      true
    );
  });

  it('CvTabs - test slotted tab buttons to match snapshot', async () => {
    const wrapper = await mount(CvTabs, {
      props: {},
      slots: {
        default: ContentTabs,
        [Tab1Label]: 'Hello <strong style="color: red;">(*)</strong>',
        [Tab2Label]: 'Origin<strong style="color: orange;">(!)</strong>',
      },
      attrs: {
        class: 'ABC-class-123',
        'aria-label': ariaLabel,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
