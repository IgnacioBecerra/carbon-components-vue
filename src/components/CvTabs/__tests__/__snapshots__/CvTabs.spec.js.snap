// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvTabs CvTabs - test slotted tab buttons to match snapshot 1`] = `
<div class="cv-tabs ABC-class-123" style="width: 100%;" aria-label="ABC-aria-label-123">
  <div data-tabs="" class="cv-tab bx--tabs--scrollable ABC-class-123" role="navigation" aria-label="ABC-aria-label-123"><button aria-hidden="true" aria-label="scroll left" class="bx--tab--overflow-nav-button--hidden" tabindex="-1" type="button"><svg focusable="false" preserveAspectRatio="xMidYMid meet" xmlns="http://www.w3.org/2000/svg" fill="currentColor" width="16" height="16" viewBox="0 0 16 16" aria-hidden="true">
        <path d="M5 8L10 3 10.7 3.7 6.4 8 10.7 12.3 10 13z"></path>
      </svg></button>
    <!--v-if-->
    <ul class="bx--tabs--scrollable__nav" role="tablist">
      <li class="cv-tabs-button  bx--tabs--scrollable__nav-item bx--tabs__nav-item--selected bx--tabs--scrollable__nav-item--selected" role="presentation"><button id="tab-1-link" class="bx--tabs--scrollable__nav-link" role="tab" aria-controls="tab-1" aria-disabled="false" aria-selected="true" tabindex="0" type="button">Hello <strong style="color: red;">(*)</strong></button></li>
      <li class="cv-tabs-button  bx--tabs--scrollable__nav-item" role="presentation"><button id="tab-2-link" class="bx--tabs--scrollable__nav-link" role="tab" aria-controls="tab-2" aria-disabled="false" aria-selected="false" tabindex="-1" type="button">Origin<strong style="color: orange;">(!)</strong></button></li>
      <li class="cv-tabs-button  bx--tabs--scrollable__nav-item" role="presentation"><button id="tab-3-link" class="bx--tabs--scrollable__nav-link" role="tab" aria-controls="tab-3" aria-disabled="false" aria-selected="false" tabindex="-1" type="button">Tab 3 label</button></li>
    </ul>
    <!--v-if--><button aria-hidden="true" aria-label="scroll right" class="bx--tab--overflow-nav-button--hidden" tabindex="-1" type="button"><svg focusable="false" preserveAspectRatio="xMidYMid meet" xmlns="http://www.w3.org/2000/svg" fill="currentColor" width="16" height="16" viewBox="0 0 16 16" aria-hidden="true">
        <path d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"></path>
      </svg></button>
  </div>
  <div class="cv-tabs__panels">
    <div id="tab-1" class="cv-tab bx--tab-content" role="tabpanel" aria-labelledby="tab-1-link" aria-hidden="false">Tab 1 content</div>
    <div id="tab-2" class="cv-tab bx--tab-content" role="tabpanel" aria-labelledby="tab-2-link" aria-hidden="true" hidden="">Tab 2 content</div>
    <div id="tab-3" class="cv-tab bx--tab-content" role="tabpanel" aria-labelledby="tab-3-link" aria-hidden="true" hidden="">Tab 3 content</div>
  </div>
</div>
`;
